index.bs

```diff
@@ -1,164 +1,194 @@
 
-The `DW_OP_WASM_location` with code `0xED` can be added to the DWARF expression language. This operation will have two operands.
+2. Memory location descriptions (see 2.6.1.1.2 of [[DWARF]]) are used when a value is located at some address in memory.
 
-The first operand defines the type of the WebAssembly location. It is encoded as ULEB128 constant.
+3. Register location descriptions (see 2.6.1.1.3 of [[DWARF]]) are used when a value is located in a register.
 
-The second operand is a location item index and it is encoded as ULEB128 constant.
+4. Implicit location descriptions (see 2.6.1.1.4 of [[DWARF]]) are used when a value does not have any runtime representation, but has a known value anyways.
+
+Each of these location descriptions are applicable to values in WebAssembly, and may be used as they normally are, except for the third: register location descriptions. WebAssembly does not have registers per se. Instead, it has three distinct kinds of virtual registers (globals, locals, and the operand stack) and may use up to 2<sup>32</sup> - 1 instances of each virtual register.
+</div>
+
+When a program object's value is stored in a WebAssembly local, global, or on the operand stack it must be encoded as an offset from a well-known register number with the <code>DW_OP_breg<sub>n</sub></code> DWARF expression operation (see 2.5.1.2 of [[DWARF]]).
 
 <table class=index>
-  <caption>WebAssembly Location</caption>
+  <caption>Summary of encoding WebAssembly-specific constructs as DWARF registers and offsets</caption>
   <thead>
     <tr>
-      <th>Type</th><th>Description</th><th>Second Parameter</th>
+      <th>DWARF Register Number</th><th>WebAssembly Construct</th><th>Meaning of offset i</th>
     </tr>
   </thead>
   <tbody>
     <tr>
-      <td>1</td><td>Local or Argument</td><td>The index of a WebAssembly function  local/parameter</td>
+      <td>1</td><td>Local</td><td>The i<sup>th</sup> local of the function.</td>
     </tr>
     <tr>
-      <td>2</td><td>Global</td><td>The index of a WebAssembly module global</td>
+      <td>2</td><td>Global</td><td>The i<sup>th</sup> global of the module.</td>
     </tr>
     <tr>
-      <td>3</td><td>Operand Stack Item</td><td>The depth of the item (0 - points the bottom of the stack)</td>
+      <td>3</td><td>Operand Stack</td><td>The i<sup>th</sup> item on the operand stack.</td>
     </tr>
   </tbody>
 </table>
 
+<div class="note">
+Note: An alternative approach to encoding locals, globals, and the operand stack would be to pack them into DWARF registers directly. If there are L locals, and G globals, then DWARF registers 0 <= g < G would reference each global, DWARF registers G <= l < G+L would reference each local, and DWARF registers s >= G+L would reference values on the operand stack. *TODO: what are the downsides of this approach again? Why aren't we doing this? Should explain that here.*
```

Comment:
I don't really see a reason for such encoding scheme. As I mentioned above, the DWARF spec defines a specific meaning to the operations, and locals, global and operand stack do not fit this meaning (that's in addition to implementation details in llvm/lldb)

Since there is no intent or ability to use this DWARF format as is in native debuggers or existing tooling, mixing DWARF defined meaning of operands with WebAssembly one might raise concern from standardization and tools maintaining point of view: it is easier to add appendix/extensions to DWARF spec rather than change narrative of e.g. `DW_OP_bregn`. What do you think?

Reply:
Agreed that it is easier and more directly expresses intent to add extensions (especially since DWARF goes out of its way to be extensible). And that is something that could be documented right here, for any future readers who are questioning why decisions went one way or another.